src/benchmarks/micro/corefx/System.Text.Json/Utf8JsonWriter/ArrayBufferWriter.cs

```diff
@@ -0,0 +1,136 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Buffers;
+using System.Diagnostics;
+
+namespace System.Text.Json
+{
+    internal sealed class ArrayBufferWriter<T> : IBufferWriter<T>
+    {
+        private T[] _buffer;
+        private int _index;
+
+        private const int MinimumBufferSize = 256;
+
+        public ArrayBufferWriter()
+        {
+            _buffer = new T[MinimumBufferSize];
+            _index = 0;
+        }
+
+        public ArrayBufferWriter(int initialCapacity)
+        {
+            if (initialCapacity <= 0)
+                throw new ArgumentException(nameof(initialCapacity));
+
+            _buffer = new T[initialCapacity];
+            _index = 0;
+        }
+
+        public ReadOnlyMemory<T> WrittenMemory
+        {
+            get
+            {
+                return _buffer.AsMemory(0, _index);
+            }
+        }
+
+        public int WrittenCount
+        {
+            get
+            {
+                return _index;
+            }
+        }
+
+        public int Capacity
+        {
+            get
+            {
+                return _buffer.Length;
+            }
+        }
+
+        public int FreeCapacity
+        {
+            get
+            {
+                return _buffer.Length - _index;
+            }
+        }
+
+        public void Clear()
+        {
+            ClearHelper();
+        }
+
+        private void ClearHelper()
+        {
+            Debug.Assert(_buffer != null);
+
+            _buffer.AsSpan(0, _index).Clear();
+            _index = 0;
+        }
+
+        public void Advance(int count)
+        {
+            if (count < 0)
+                throw new ArgumentException(nameof(count));
+
+            if (_index > _buffer.Length - count)
+                ThrowInvalidOperationException(_buffer.Length);
+
+            _index += count;
+        }
+
+        public Memory<T> GetMemory(int sizeHint = 0)
```

Comment:
For simplicity of the benchmarks, which run in a controlled scenario, we can allocate enough memory up front for all cases, and throw if the requested buffer size is not available.
From the docs:
```
Remarks
There is no guarantee that successive calls will return the same buffer or the same-sized buffer.
This method must never return Span<T>.Empty, but it can throw if the requested buffer size is not available.
You must request a new buffer after calling Advance to continue writing more data; you cannot write to a previously acquired buffer.
```

Reply:
That does not capture the "real" usage of the JsonWriter. The ArrayBufferWriter, as it is defined here, will be publicly available/built-in as part of .NET Core 3.0. Generally, users will not be able to allocate enough memory up front.

Reply:
For the "real" usage of the `JsonWriter` you need real world code, this is measuring components and of the new feature without writing unit tests or testing other .NET component.

Reply:
I would rather not make this change since adding new benchmark tests become difficult here. The contributor would have to figure out what the required capacity needs to be (i.e. run it, see WrittenCount, and then set initialCapacity of ArrayBufferWriter to max count seen for all cases).

Given this type will be built-in soon, and we will be using that instead, I don't see much benefit of modifying the implementation like this atm. Let me know if you disagree.

In theory, I could change ArrayBufferWriter to be as follows, since nothing else is required, but I don't really see a benefit to that.

```C#

using System.Buffers;

namespace System.Text.Json.Tests
{
    internal sealed class ArrayBufferWriter<T> : IBufferWriter<T>
    {
        private readonly T[] _buffer;
        private int _index;

        public ArrayBufferWriter(int initialCapacity)
        {
            if (initialCapacity <= 0)
                throw new ArgumentException(nameof(initialCapacity));

            _buffer = new T[initialCapacity];
            _index = 0;
        }

        public void Clear()
        {
            _buffer.AsSpan(0, _index).Clear();
            _index = 0;
        }

        public void Advance(int count)
        {
            if (count < 0)
                throw new ArgumentException(nameof(count));
            if (_index > _buffer.Length - count)
                throw new InvalidOperationException($"Cannot advance past the end of the buffer, which has a size of {_buffer.Length}.");

            _index += count;
        }

        public Memory<T> GetMemory(int sizeHint = 0)
        {
            if (sizeHint < 0)
                throw new ArgumentException(nameof(sizeHint));
            if (sizeHint > _buffer.Length - _index)
                throw new InvalidOperationException($"The initial buffer is not large enough to continue. It has a size of {_buffer.Length}.");

            return _buffer.AsMemory(_index);
        }

        public Span<T> GetSpan(int sizeHint = 0)
        {
            if (sizeHint < 0)
                throw new ArgumentException(nameof(sizeHint));
            if (sizeHint > _buffer.Length - _index)
                throw new InvalidOperationException($"The initial buffer is not large enough to continue. It has a size of {_buffer.Length}.");

            return _buffer.AsSpan(_index);
        }
    }
}

```